File: src/riego/sensores/HumedadReaderTask.java
package riego.sensores;

public class HumedadReaderTask implements Runnable {
    private volatile double ultimaHumedad = Double.NaN;
    private volatile boolean ejecutando = true;

    @Override
    public void run() {
        while (ejecutando) {
            try {
                ultimaHumedad = leerSensor();
                System.out.printf("[Humedad] %.2f %% %n", ultimaHumedad);
                Thread.sleep(3000); // simula muestreo
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
                break;
            }
        }
    }

    private double leerSensor() {
        return Math.random() * 100; // entre 0% y 100%
    }

    public double getUltimaHumedad() {
        return ultimaHumedad;
    }

    public void detener() {
        ejecutando = false;
    }
}

File: src/riego/sensores/humedad_reader_task.py
import threading, random, time
from src.patrones.observer.observable import Observable

class HumedadReaderTask(threading.Thread, Observable[float]):
    def __init__(self):
        threading.Thread.__init__(self, daemon=True)
        Observable.__init__(self)
        self._detenido = threading.Event()

    def run(self):
        while not self._detenido.is_set():
            humedad = random.uniform(0, 100)       # genera humedad aleatoria
            self.notificar_observadores(humedad)   # avisa a los observadores
            time.sleep(3.0)                        # espera 3 segundos

    def detener(self):
        self._detenido.set()
